vpr/src/base/xdc_constraints.cpp

```diff
@@ -0,0 +1,244 @@
+            package_pin_region.y,
+            package_pin_region.x + package_pin_region.w - 1,
+            package_pin_region.y + package_pin_region.h - 1
+        );
+        r.set_sub_tile(package_pin_region.subtile);
+        PartitionRegion pr;
+        pr.add_to_part_region(r);
+        part.set_part_region(pr);
+        PartitionId part_id = this->constraints.add_partition(part);
+
+        /* Constrain the AtomBlock */
+
+        this->constraints.add_constrained_atom(port_block, part_id);
+
+        VTR_LOG("XDC: Assigned PACKAGE_PIN %s = [%d, %d, %d, %d].%d\n",
+                pin_name, package_pin_region.x, package_pin_region.y,
+                package_pin_region.w, package_pin_region.h, package_pin_region.subtile);
+
+        return this->_ret_ok();
+    }
+
+    int _set_property_iostandard(Tcl_Obj* tcl_io_standard, Tcl_Obj* tcl_port) {
+        const char* io_standard = Tcl_GetString(tcl_io_standard);
+        if (tcl_io_standard == nullptr)
+            return this->_ret_error("set_property: iostandard should be a string.");
+        
+        auto* port = tcl_obj_getptr<AtomPortId>(tcl_port);
+        if (port == nullptr)
+            return this->_ret_error("set_property: port_name of IOSTANDARD should have a `port` type.");
+        
+        return this->_do_set_property_iostandard(io_standard, *port);
+    }
+
+    int _do_set_property_iostandard(const char* io_standard, AtomPortId port) {
+        AtomBlockId block = this->netlist.port_block(port);
+        this->netlist.set_block_param(block, "IOSTANDARD", io_standard);
+        VTR_LOG("XDC: Assigned IOSTANDARD %s = %s\n", this->netlist.block_name(block).c_str(), io_standard);
+    }
+
+    int _do_get_ports(const char* pin_name) {
+        AtomBlockId pin_block = this->netlist.find_block(pin_name);
+
+        if (pin_block == AtomBlockId::INVALID()) {
+            std::string available_ports;
+            const auto& pins = this->netlist.blocks();
+            int pins_left = pins.size();
+            for (auto& pin_block_ : pins) {
+                auto ports = this->netlist.block_ports(pin_block_);
+                auto port_it = this->netlist.block_ports(pin_block_).begin();
+                if (port_it != ports.end()) {
+                    available_ports += this->netlist.block_name(pin_block_);
+                    if (pins_left != 1)
+                        available_ports += ", ";
+                }
+                pins_left--;
+            }
+
+            return this->_ret_error("set_property: Can't find ports for `" +
+                                    std::string(pin_name) + "`. Available blocks: " +
+                                    available_ports);
+        }
+        
+        auto ports = this->netlist.block_ports(pin_block);
+        auto port_it = this->netlist.block_ports(pin_block).begin();
+        if (port_it == ports.end())
+            return this->_ret_error("set_property: '" + std::string(pin_name) + "' - no ports found.");
```

Comment:
The error should probably be corresponding to `get_ports` rather than `set_property`.